test/metabase/lib/datascript_test.cljc

```diff
@@ -0,0 +1,1100 @@
+                              ;; Scalars - just compare them.
+                              (=?/=?-diff exp-v act-v))]
+               :when diff]
+           {exp-k diff})))
+
+;; We know in this function that every exp has at least one match! This is only to find that there's no possible
+;; matching.
+(defn- =?-diff-sets-matching [act-set [[exp {:keys [matches]}] & pairs]]
+  (let [available (filter act-set matches)]
+    (cond
+      (empty? available) #{exp}
+      (empty? pairs)     nil    ;; Successful match!
+      :else (let [tails (for [act available]
+                          (=?-diff-sets-matching (disj act-set act) pairs))]
+              (if (some nil? tails)
+                nil ; Successfully found a matching.
+                ;; No matches. Each one returns a set of disappointed exps. Return that union.
+                (reduce set/union nil tails))))))
+
+(#?@(:clj  (methodical/defmethod =?/=?-diff [clojure.lang.Sequential   clojure.lang.IPersistentSet])
+     :cljs (defmethod            =?/=?-diff [:dispatch-type/sequential :dispatch-type/set]))
+  [exps act-set]
+  ;; Each exp must be matched with something in the act-set. The order doesn't matter, but there must exist a matching.
+  ;; If there are no matching for an exp, we want to show... all the diffs? Just the smallest one?
+  (let [;; Results is a parallel list to exps: [exp {:diffs {act diff ...}, :matched true}]
+        results (for [exp exps
+                      :let [diffs (into {} (map (fn [act]
+                                                  [act (=?/=?-diff exp act)]))
+                                        act-set)]]
+                  [exp {:diffs   (m/filter-vals some? diffs)
+                        :matches (set (keys (m/filter-vals nil? diffs)))}])
+        unmatched (keep (fn [[_exp {:keys [diffs matches]}]]
+                          (when (empty? matches)
+                            (set (vals diffs))))
+                        results)]
+    (if (seq unmatched)
+      ;; Returns unmatched with a list of [exp #{diffs...}] pairs.
+      ;; TODO: Probably too long?
+      (list* 'unmatched unmatched)
+      ;; Otherwise try to match everything up consistently.
+      (if-let [diff (=?-diff-sets-matching act-set results)]
+        (list 'no-matching-among diff)
+        nil))))
+
+(def ^:private sample-5-stages
+  {:db/id             -10
+   :mbql/database     [:metadata.database/id 1]
+
+   :mbql.stage/_query [{:db/id              -1}
+                       {:db/id              -2
+                        :mbql.stage/source  -1}
+                       {:db/id              -3
+                        :mbql.stage/source  -2}
+                       {:db/id              -4
+                        :mbql.stage/source  -3}
+                       {:db/id              -5
+                        :mbql.stage/source  -4}]
+   :mbql.query/stage0 -1
+   :mbql.query/stage  -5})
+
+(deftest ^:parallel stage-rules-test
```

Comment:
Testing the `d/q` querying "rules" that find a stage by its number. That's one of the first things I wrote, but in hindsight it doesn't really make sense. `query-stage` would be easier and probably faster just walking the Entity maps.